lib/Mojolicious/Validator.pm

```diff
@@ -0,0 +1,221 @@
+
+  return $self;
+}
+
+sub validation { Mojolicious::Validator::Validation->new(validator => shift) }
+
+sub _comma_separated { defined $_[2] ? split(/\s*,\s*/, $_[2], -1) : undef }
+
+sub _equal_to {
+  my ($v, $name, $value, $to) = @_;
+  return 1 unless defined(my $other = $v->input->{$to});
+  return $value ne $other;
+}
+
+sub _in {
+  my ($v, $name, $value) = (shift, shift, shift);
+  $value eq $_ && return undef for @_;
+  return 1;
+}
+
+sub _not_empty { length $_[2] ? $_[2] : () }
+
+sub _num {
+  my ($v, $name, $value, $min, $max) = @_;
+  return 1 if $value !~ /^-?[0-9]+$/;
+  return defined $min && $min > $value || defined $max && $max < $value;
+}
+
+sub _size {
+  my ($v, $name, $value, $min, $max) = @_;
+  my $len = ref $value ? $value->size : length $value;
+  return (defined $min && $len < $min) || (defined $max && $len > $max);
+}
+
+sub _trim { defined $_[2] ? trim $_[2] : undef }
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojolicious::Validator - Validate values
+
+=head1 SYNOPSIS
+
+  use Mojolicious::Validator;
+
+  my $validator = Mojolicious::Validator->new;
+  my $v = $validator->validation;
+  $v->input({foo => 'bar'});
+  $v->required('foo')->like(qr/ar$/);
+  say $v->param('foo');
+
+=head1 DESCRIPTION
+
+L<Mojolicious::Validator> validates values for L<Mojolicious>.
+
+=head1 CHECKS
+
+These validation checks are available by default.
+
+=head2 equal_to
+
+  $v = $v->equal_to('foo');
+
+String value needs to be equal to the value of another field.
+
+=head2 in
+
+  $v = $v->in('foo', 'bar', 'baz');
+
+String value needs to match one of the values in the list.
+
+=head2 like
+
+  $v = $v->like(qr/^[A-Z]/);
+
+String value needs to match the regular expression.
+
+=head2 num
+
+  $v = $v->num;
+  $v = $v->num(2, 5);
+  $v = $v->num(-3, 7);
+  $v = $v->num(2, undef);
+  $v = $v->num(undef, 5);
+
+String value needs to be a non-fractional number (positive or negative) and if provided in the given range.
+
+=head2 size
+
+  $v = $v->size(2, 5);
+  $v = $v->size(2, undef);
+  $v = $v->size(undef, 5);
+
+String value length or size of L<Mojo::Upload> object in bytes needs to be between these two values.
+
+=head2 upload
+
+  $v = $v->upload;
+
+Value needs to be a L<Mojo::Upload> object, representing a file upload.
+
+=head1 FILTERS
+
+These filters are available by default.
+
+=head2 comma_separated
+
+  $v = $v->optional('foo', 'comma_separated');
+
+Split string of comma separated values into separate values.
+
+=head2 not_empty
+
+  $v = $v->optional('foo', 'not_empty');
+
+Remove empty string values and treat them as if they had not been submitted.
+
+=head2 trim
+
+  $v = $v->optional('foo', 'trim');
+
+Trim whitespace characters from both ends of string value with L<Mojo::Util/"trim">.
+
+=head1 ATTRIBUTES
+
+L<Mojolicious::Validator> implements the following attributes.
+
+=head2 checks
+
+  my $checks = $validator->checks;
+  $validator = $validator->checks({size => sub ($v, $name, $value, @args) {...}});
```

Comment:
It seems very important to document here that the expected return value is false for a check that passes, and a true value that is accessible via the ->error method for a check that fails.